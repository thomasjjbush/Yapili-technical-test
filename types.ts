import { ReactNode } from 'react';
import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import HtmlWebPackPlugin from 'html-webpack-plugin';
import { Json } from 'enzyme-to-json';
import { MockStoreEnhanced } from 'redux-mock-store';

export type Action = ThunkAction<Promise<AnyAction | void>, Store, unknown, AnyAction>;

export interface AspectRatioProps {
    children: ReactNode;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '16:9' = 56.25,
    '4:3' = 75,
}

export interface Character {
    comics?: CharecterCollection;
    description?: string;
    error?: boolean;
    events?: CharecterCollection;
    id?: number;
    modified?: string;
    name?: string;
    resourceURI?: string;
    series?: CharecterCollection;
    stories?: CharecterCollection;
    thumbnail?: {
        extension: string;
        path: string;
    };
}

export enum CharacterActions {
    CLEARED = 'character/CLEARED',
    ERROR = 'character/ERROR',
    LOADED = 'character/LOADED',
}

interface CharecterCollection {
    available?: number;
    collectionURI?: string;
    items: Resource[];
    returned?: number;
}

export enum Dimensions {
    LANDSCAPE = 'landscape',
    STANDARD = 'standard',
}

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export interface FallbackImageProps {
    src: string;
    width?: string;
    height?: string;
}

export interface Home extends Partial<MarvelResponse> {
    error?: boolean;
    results?: Character[];
}

export enum HomeActions {
    ERROR = 'home/ERROR',
    LOADED = 'home/LOADED',
    LOADING = 'home/LOADING',
}

export interface ImageProps {
    dimension?: Dimensions;
    extension: string;
    height?: string;
    path: string;
    width?: string;
}

export interface MarvelResponse {
    count: number;
    limit: number;
    offset: number;
    results: any[];
    total: number;
}

export type MockStore = MockStoreEnhanced<Partial<Store>, Dispatch>;

export interface Options {
    [key: string]: string | number;
    apikey?: string;
    limit?: number;
    offset?: number;
}

export interface PaginatedButtonsProps {
    currentPage: number;
    numberOfButtons: number;
    numberOfPages: number;
    skip?: number;
    to: (i: number) => Partial<Location>;
}

export interface Resource {
    resourceURI: string;
    name: string;
    type?: string;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export interface Store {
    character: Character;
    home: Home;
}

export type StyledProps<Props = {}> = { theme: Theme } & Props;

export interface Theme {
    breakpoints: {
        desktop: number;
        tabletLandscape: number;
        tabletPortrait: number;
    };
    colors: {
        darkGrey: string;
        marvelRed: string;
        mediumGrey: string;
    };
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: HtmlWebPackPlugin[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
