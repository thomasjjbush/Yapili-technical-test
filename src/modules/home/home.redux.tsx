import { AnyAction } from 'redux';
import { Action, Home, HomeActions } from '../../../types';
import { createAction } from '../../utils/action/action';
import { db } from '../../utils/db/db';

const loadingResults = [...new Array(20)].map((_, i) => ({ id: i, name: '...' }));

export const loadCharacters = (multiplier: number): Action => async (dispatch, getState) => {
    const offset = multiplier * 20;
    if (getState().home.offset !== offset) dispatch(createAction(HomeActions.LOADING));

    try {
        const payload = await db('characters', { offset });
        dispatch(createAction(HomeActions.LOADED, payload));
    } catch (err) {
        dispatch(createAction(HomeActions.ERROR));
    }
};

export const homeReducer = (state = {}, { payload, type }: AnyAction): Home => {
    switch (type) {
        case HomeActions.ERROR:
            return { ...state, error: true };
        case HomeActions.LOADED:
            return { ...state, ...(payload as Home) };
        case HomeActions.LOADING:
            return { ...state, results: loadingResults };
        default:
            return state;
    }
};
