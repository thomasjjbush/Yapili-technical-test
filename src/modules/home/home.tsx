import React, { FC, ReactElement, useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useLocation } from 'react-router-dom';
import styled from 'styled-components';
import { loadCharacters } from './home.redux';
import { AspectRatio } from '../../components/aspect-ratio/aspect-ratio';
import { PaginatedButtons } from '../../components/paginated-buttons/paginated-buttons';
import { Image } from '../../components/image/image';
import { AspectRatios, Home as HomeState, Store, StyledProps } from '../../../types';
import { Grid } from '../../components/grid/grid';

const Character = styled(Link)<StyledProps>`
    background-color: ${({ theme }): string => theme.colors.darkGrey};
    border-bottom: solid 5px ${({ theme }): string => theme.colors.marvelRed};
    overflow: hidden;
`;

const CharacterInfo = styled.div`
    overflow: hidden;
    padding: 20px;
`;

const Description = styled.p`
    margin: 20px 0 0;
`;

const Name = styled.h1`
    margin: 0;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
`;

export const Home: FC = (): ReactElement => {
    const dispatch = useDispatch();
    const { results, total } = useSelector<Store, HomeState>((state) => state.home);
    const page = Math.max(Number(new URLSearchParams(useLocation().search).get('page')), 1);

    const to = useCallback((i: number) => ({ search: `page=${i}` }), []);

    useEffect(() => {
        window.scrollTo(0, 0);
        dispatch(loadCharacters(page - 1));
    }, [page]);

    return (
        <div>
            <Grid desktop={4} gap={30} mobile={1} tablet={2}>
                {results?.map(({ description, id, name, thumbnail }) => (
                    <Character to={`/character/${id}`} key={id}>
                        <AspectRatio ratio={AspectRatios['1:1']}>
                            {thumbnail && (
                                <Image
                                    extension={thumbnail.extension}
                                    height="100%"
                                    path={thumbnail.path}
                                    width="100%"
                                />
                            )}
                        </AspectRatio>
                        <CharacterInfo>
                            <Name>{name}</Name>
                            {description && <Description>{description}</Description>}
                        </CharacterInfo>
                    </Character>
                ))}
            </Grid>
            {total && (
                <PaginatedButtons
                    currentPage={page}
                    numberOfButtons={5}
                    numberOfPages={Math.floor(total / 20)}
                    skip={5}
                    to={to}
                />
            )}
        </div>
    );
};
