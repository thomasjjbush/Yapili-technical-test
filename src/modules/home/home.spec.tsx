import React, { ReactElement } from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { Home } from './home';
import { Provider } from 'react-redux';
import { BrowserRouter, Route } from 'react-router-dom';
import { mockStore, mockTheme } from '../../../test/jest.utils';
import { loadCharacters } from './home.redux';
import { Home as HomeT } from '../../../types';
import { ThemeProvider } from 'styled-components';

interface P {
    children: ReactElement;
}

const spy = jest.spyOn(window, 'scrollTo').mockImplementation();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    Link: (props: P) => props.children || 'Link',
}));
jest.mock('./home.redux', () => ({ loadCharacters: jest.fn().mockReturnValue({ type: 'MOCK_ACTION' }) }));
jest.mock('../../components/image/image', () => ({ Image: () => 'Image' }));
jest.mock('../../components/grid/grid', () => ({ Grid: (props: P) => props.children || 'Grid' }));
jest.mock('../../components/aspect-ratio/aspect-ratio', () => ({ AspectRatio: (props: P) => props.children }));
jest.mock('../../components/paginated-buttons/paginated-buttons', () => ({
    PaginatedButtons: () => 'PaginatedButtons',
}));

const ConnectedComponent = (home?: HomeT): ReactWrapper =>
    mount(
        <Provider store={mockStore({ home })}>
            <ThemeProvider theme={mockTheme}>
                <BrowserRouter>
                    <Route path="/">
                        <Home />
                    </Route>
                </BrowserRouter>
            </ThemeProvider>
        </Provider>,
    );

describe('Character', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should invoke load character and scroll top', () => {
        ConnectedComponent();
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(0, 0);
        expect(loadCharacters).toHaveBeenCalledTimes(1);
        expect(loadCharacters).toHaveBeenCalledWith(0);
    });

    it('should match snapshot', () => {
        expect(
            ConnectedComponent({
                total: 10,
                offset: 0,
                results: [
                    {
                        description: 'description',
                        id: 55,
                        name: 'name',
                        thumbnail: { path: 'path', extension: 'ext' },
                    },
                ],
            }).find(Home),
        ).toMatchSnapshot();
    });
});
