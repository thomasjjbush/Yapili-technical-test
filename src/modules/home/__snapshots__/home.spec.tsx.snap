// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character should match snapshot 1`] = `
<Home>
  <div>
    <Grid
      desktop={4}
      gap={30}
      mobile={1}
      tablet={2}
    >
      <Styled(Link)
        key="55"
        styles="
    background-color: theme.colors.darkGrey;
    border-bottom: solid 5px theme.colors.marvelRed;
    overflow: hidden;
"
        to="/character/55"
      >
        <Link
          className="sc-bdfBwQ bWCLJs"
          to="/character/55"
        >
          <AspectRatio
            ratio={100}
          >
            <Image
              extension="ext"
              height="100%"
              path="path"
              width="100%"
            >
              Image
            </Image>
          </AspectRatio>
          <styled.div
            styles="
    overflow: hidden;
    padding: 20px;
"
          >
            <div
              className="sc-gsTCUz zwlLb"
            >
              <styled.h1
                styles="
    margin: 0;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
"
              >
                <h1
                  className="sc-hKgILt jouoAZ"
                >
                  name
                </h1>
              </styled.h1>
              <styled.p
                styles="
    margin: 20px 0 0;
"
              >
                <p
                  className="sc-dlfnbm dutTAS"
                >
                  description
                </p>
              </styled.p>
            </div>
          </styled.div>
        </Link>
      </Styled(Link)>
    </Grid>
    <PaginatedButtons
      currentPage={1}
      numberOfButtons={5}
      numberOfPages={0}
      skip={5}
      to={[Function]}
    >
      PaginatedButtons
    </PaginatedButtons>
  </div>
</Home>
`;
